pkg/asset/store.go

```diff
@@ -162,92 +164,120 @@ func (s *StoreImpl) save() error {
 // fetch populates the given asset, generating it and its dependencies if
 // necessary, and returns whether or not the asset had to be regenerated and
 // any errors.
-func (s *StoreImpl) fetch(asset Asset, indent string) (bool, error) {
+func (s *StoreImpl) fetch(asset Asset, indent string) error {
 	logrus.Debugf("%sFetching %q...", indent, asset.Name())
 
 	// Return immediately if the asset is found in the cache,
 	// this is because we are doing a depth-first-search, it's guaranteed
 	// that we always fetch the parent before children, so we don't need
 	// to worry about invalidating anything in the cache.
-	storedAsset, ok := s.assets[reflect.TypeOf(asset)]
-	if ok {
+	if storedAsset, ok := s.assets[reflect.TypeOf(asset)]; ok {
 		logrus.Debugf("%sReusing previously-fetched %q", indent, asset.Name())
-		reflect.ValueOf(asset).Elem().Set(reflect.ValueOf(storedAsset.asset).Elem())
-		return storedAsset.dirty, nil
+		reflect.ValueOf(asset).Elem().Set(reflect.ValueOf(storedAsset).Elem())
+		return nil
+	}
+
+	// Load from on-disk any assets in the graph starting at this asset.
+	onDiskAssetState, err := s.loadFromOnDisk(asset, indent)
+	if err != nil {
+		return err
+	}
+
+	// If the asset is on-disk, try to use the on-disk asset.
+	if onDiskAssetState.present {
```

Comment:
can we move this to `Fetch`, it looks like we only need to call this once per requested asset.